File: src/main/java/it/gabrieletondi/telldontaskkata/useCase/OrderShipmentUseCase.java
package it.gabrieletondi.telldontaskkata.useCase;

import it.gabrieletondi.telldontaskkata.domain.Order;
import it.gabrieletondi.telldontaskkata.domain.OrderStatus;
import it.gabrieletondi.telldontaskkata.repository.OrderRepository;
import it.gabrieletondi.telldontaskkata.service.ShipmentService;

import java.util.Optional;

import static it.gabrieletondi.telldontaskkata.domain.OrderStatus.CREATED;
import static it.gabrieletondi.telldontaskkata.domain.OrderStatus.REJECTED;
import static it.gabrieletondi.telldontaskkata.domain.OrderStatus.SHIPPED;

public class OrderShipmentUseCase {
    private final OrderRepository orderRepository;
    private final ShipmentService shipmentService;

    public OrderShipmentUseCase(OrderRepository orderRepository, ShipmentService shipmentService) {
        this.orderRepository = orderRepository;
        this.shipmentService = shipmentService;
    }

    public void run(OrderShipmentRequest request) {
        final Order order = Optional.ofNullable(orderRepository.getById(request.getOrderId()))
                .orElseThrow(() -> new RuntimeException("Order Id does not exist"));

        validateOrder(order);

        shipmentService.ship(order);

        order.setStatus(OrderStatus.SHIPPED);
        orderRepository.save(order);
    }

    public void validateOrder(Order order) {
        if (isOrderShippable(order)) {
            throw new OrderCannotBeShippedException();
        }

        if (order.isOrderStatus(SHIPPED)) {
            throw new OrderCannotBeShippedTwiceException();
        }
    }

    public boolean isOrderShippable(Order order){
        return order.isOrderStatus(CREATED) || order.isOrderStatus(REJECTED);
    }
}

File: kotlin/src/main/java/it/gabrieletondi/telldontaskkata/useCase/OrderShipmentUseCase.kt
package it.gabrieletondi.telldontaskkata.useCase

import it.gabrieletondi.telldontaskkata.domain.OrderStatus
import it.gabrieletondi.telldontaskkata.repository.OrderRepository
import it.gabrieletondi.telldontaskkata.service.ShipmentService

class OrderShipmentUseCase(private val orderRepository: OrderRepository, private val shipmentService: ShipmentService) {
    fun run(request: OrderShipmentRequest) {
        val order = orderRepository.getById(request.orderId)
        if (order!!.status == OrderStatus.CREATED || order.status == OrderStatus.REJECTED) {
            throw OrderCannotBeShippedException()
        }
        if (order.status == OrderStatus.SHIPPED) {
            throw OrderCannotBeShippedTwiceException()
        }
        shipmentService.ship(order)
        order.status = OrderStatus.SHIPPED
        orderRepository.save(order)
    }
}